File: Hashing_and_HashMap_problems/SortThePeople.java

// Problem link : https://leetcode.com/problems/sort-the-people/description/?envType=problem-list-v2&envId=hash-table

package Hashing_and_HashMap_problems;

import java.util.Arrays;

public class SortThePeople {
    public static String[] sortPeople(String[] names, int[] heights) {
        // step 1 - create an array of indices
        Integer[] indices = new Integer[heights.length];
        for(int i=0; i<heights.length; i++){
            indices[i] = i;
        }

        // step 2 - sort indices based on heights in descending order
        Arrays.sort(indices, (a, b) -> Integer.compare(heights[b], heights[a]));

        // step 3 - create the result array using sorted indices
        String[] res = new String[names.length];
        for(int i=0; i<indices.length; i++){
            res[i] = names[indices[i]];
        }

        // step 4 - return resutant array `res`
        return res;
    }

    public static void main(String[] args) {
        String[] names = {"Mary","John","Emma"};
        int[] heights = {180, 165, 170};
        String[] res = sortPeople(names, heights);
        System.out.println(Arrays.toString(res));
    }
}

File: Hashing_and_HashMap_problems/SortThePeople.js

// Problem link : https://leetcode.com/problems/sort-the-people/description/?envType=problem-list-v2&envId=hash-table

let sortPeople1 = function(names, heights) {
    const pairs = names.map((name, index) => [name, heights[index]]);

    pairs.sort((a,b) => b[1] - a[1]);

    const res = [];
    for(let i=0; i<pairs.length; i++){
        res.push(pairs[i][0]);
    }
    
    return res;
};

let sortPeople = function(names, heights) {
    // step 1 - sort indices of the heights array based on the heights in descending order
    return heights
        // step 2 - map heights to their indices 
        .map((height, index) => index)
        // step 3 - sort indices by corresponding heights 
        .sort((a,b) => heights[b] - heights[a])
        // step 4 - map sorted indices to corresponding names
        .map((index) => names[index]);
};

// const names = ["Mary","John","Emma"], heights = [180,165,170];

const names = ["Alice","Bob","Bob"], heights = [155,185,150];
console.log(sortPeople(names, heights));
